src/main/java/net/troja/eve/esi/auth/OAuth.java

```diff
@@ -26,25 +26,29 @@
     private static final String URI_AUTHENTICATION = URI_OAUTH + "/authorize";
     private static final String URI_ACCESS_TOKEN = URI_OAUTH + "/token";
 
-    private String accessToken;
-    private long validUntil;
     private String refreshToken;
     private String clientId;
     private String clientSecret;
     private OAuth2CodeGrantFlow oAuthFlow;
-    private final Map<String, AccessTokenData> accessTokenCache = new HashMap<>();
+    private static final Map<String, AccessTokenData> ACCESS_TOKEN_CACHE = new ConcurrentHashMap<>();
 
     @Override
     public void applyToParams(final List<Pair> queryParams, final Map<String, String> headerParams) {
-        if (refreshToken != null && validUntil < System.currentTimeMillis()) {
-            try {
-                refreshToken();
-            } catch (final ProcessingException ex) {
-                // This error will be handled by ESI once the request is made
+        //Check if we need a new access token
+        synchronized(OAuth.class) { //This block is synchronized across all threads - so we don't update the access token more than once
+            AccessTokenData accessTokenData = ACCESS_TOKEN_CACHE.get(getAuthKey());
+            if (refreshToken != null  && (accessTokenData == null || accessTokenData.getValidUntil() < System.currentTimeMillis())) {
+                try {
+                    refreshToken();
+                } catch (final ProcessingException ex) {
+                    // This error will be handled by ESI once the request is made
+                }
             }
```

Comment:
What should happen if refreshToken is null?  Throw an error that this token must be re-authorized?  Handled somewhere else?

Reply:
If the refresh token is null, it’s expected to be a public endpoint and no auth is added

Reply:
If it is a authed endpoint and no refresh token is set, the library will throw an ApiException when calling the endpoint

Reply:
There is another case: if the access token is set, it will add the auth (no matter if the refresh token is null or not